File: partiql-ast/src/main/java/org/partiql/ast/v1/FromJoin.java
package org.partiql.ast.v1;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.partiql.ast.v1.expr.Expr;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * TODO docs, equals, hashcode
 */
@Builder(builderClassName = "Builder")
@EqualsAndHashCode(callSuper = false)
public class FromJoin extends FromTableRef {
    @NotNull
    public final FromTableRef lhs;

    @NotNull
    public final FromTableRef rhs;

    @Nullable
    public final JoinType joinType;

    @Nullable
    public final Expr condition;

    public FromJoin(@NotNull FromTableRef lhs, @NotNull FromTableRef rhs, @Nullable JoinType joinType, @Nullable Expr condition) {
        this.lhs = lhs;
        this.rhs = rhs;
        this.joinType = joinType;
        this.condition = condition;
    }

    @NotNull
    @Override
    public Collection<AstNode> children() {
        List<AstNode> kids = new ArrayList<>();
        kids.add(lhs);
        kids.add(rhs);
        if (condition != null) {
            kids.add(condition);
        }
        return kids;
    }

    @Override
    public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
        return visitor.visitFromJoin(this, ctx);
    }
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/FromJoin.kt
package org.partiql.ast.v1

import org.partiql.ast.v1.expr.Expr

/**
 * TODO docs, equals, hashcode
 */
public class FromJoin(
    @JvmField
    public var lhs: From,
    @JvmField
    public var rhs: From,
    @JvmField
    public var type: Type?,
    @JvmField
    public var condition: Expr?,
) : From() {
    public override fun children(): Collection<AstNode> {
        val kids = mutableListOf<AstNode?>()
        kids.add(lhs)
        kids.add(rhs)
        condition?.let { kids.add(it) }
        return kids.filterNotNull()
    }

    public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R =
        visitor.visitFromJoin(this, ctx)

    /**
     * TODO docs, equals, hashcode
     */
    public enum class Type {
        INNER,
        LEFT,
        LEFT_OUTER,
        RIGHT,
        RIGHT_OUTER,
        FULL,
        FULL_OUTER,
        CROSS,
        COMMA,
        OTHER,
    }
}
